bundles/org.openhab.binding.iotawatt/src/main/java/org/openhab/binding/iotawatt/internal/service/FetchDataService.java

```diff
@@ -0,0 +1,145 @@
+            if (statusResponse.isPresent()) {
+                thingHandler.updateStatus(ThingStatus.ONLINE);
+                updateChannels(statusResponse.get());
+            } else {
+                thingHandler.updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR);
+            }
+        } catch (InterruptedException e) {
+            thingHandler.updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.NOT_YET_READY);
+        } catch (TimeoutException e) {
+            thingHandler.updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR);
+        } catch (URISyntaxException e) {
+            thingHandler.updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR, getErrorMessage(e));
+        } catch (ExecutionException e) {
+            logger.debug("Error on getting data from IoTaWatt {}", nonNullClient.hostname);
+            thingHandler.updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.NONE, getErrorMessage(e));
+        }
+    }
+
+    @Nullable
+    private String getErrorMessage(Throwable t) {
+        final Throwable cause = t.getCause();
+        return Objects.requireNonNullElse(cause, t).getMessage();
+    }
+
+    private void updateChannels(StatusResponse statusResponse) {
+        if (statusResponse.inputs() != null) {
+            updateInputs(statusResponse.inputs());
+        }
+        if (statusResponse.outputs() != null) {
+            updateOutputs(statusResponse.outputs());
+        }
+    }
+
+    private void updateInputs(List<StatusResponse.Input> inputs) {
+        for (final StatusResponse.Input input : inputs) {
+            final int channelNumber = input.channel();
+            createAndUpdateInputChannel(channelNumber, input.watts(), IoTaWattChannelType.WATTS);
+            createAndUpdateInputChannel(channelNumber, input.vrms(), IoTaWattChannelType.VOLTAGE);
+            createAndUpdateInputChannel(channelNumber, input.hz(), IoTaWattChannelType.FREQUENCY);
+            createAndUpdateInputChannel(channelNumber, input.pf(), IoTaWattChannelType.POWER_FACTOR);
+            createAndUpdateInputChannel(channelNumber, input.phase(), IoTaWattChannelType.PHASE);
+        }
+    }
+
+    private void updateOutputs(List<StatusResponse.Output> outputs) {
+        int index = 0;
+        for (final StatusResponse.Output output : outputs) {
+            final ChannelUID channelUID = new ChannelUID(thingHandler.getThingUID(),
+                    OUTPUT_CHANNEL_ID_PREFIX + toTwoDigits(index++) + "#" + output.name());
+            final Float value = output.value();
+            final IoTaWattChannelType ioTaWattChannelType = IoTaWattChannelType.fromOutputUnits(output.units());
+            thingHandler.addChannelIfNotExists(channelUID, ioTaWattChannelType);
+            thingHandler.updateState(channelUID, new QuantityType<>(value, ioTaWattChannelType.unit));
+            // TODO removed channels are not in array anymore
```

Comment:
I'm not sure what should happen here. When new channels are detected, the binding creates new openHAB channels on the Thing. But when the channels disappear, should we delete them? In case they come back for some reason the links would be lost..
It also requires CPU power on every update to detect these, because they're not in the array any more. So we'd have to go through all existing channels to compare them to the incoming array.

Reply:
Was looking into other bindings on how they handle such situation, but could not find a good example. I think it is not necesairy to check and remove the channels on each update like you said. I cannot fully ases how dynamic these channels are, i would assume they are pretty stable once added. If not you could create some daily cleanup routine or whatever strategy is best.

Reply:
The users can add and remove channels on the IoTaWatt UI. Do you feel such a daily task needs to be implemented for the first version?

Reply:
Leave the comment in and see what happens, I don’t think this is a big issue anyway.